lib/grp.gi

```diff
@@ -783,16 +805,35 @@ InstallMethod( DerivedSeriesOfGroup,
     D := DerivedSubgroup( G );
    
     while
-      # we don't know that the group has no generators
-      (not HasGeneratorsOfGroup(S[Length(S)]) or
-	    Length(GeneratorsOfGroup(S[Length(S)]))>0) and
-      ( (not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
-	    Length(AbelianInvariants(S[Length(S)]))>0) do
+      (not HasIsTrivial(S[Length(S)]) or
+	    not IsTrivial(S[Length(S)])) and
+      (
+        (not HasIsPerfectGroup(S[Length(S)]) and
+         not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
+        (HasIsPerfectGroup(S[Length(S)]) and not IsPerfectGroup(S[Length(S)]))
+        or (HasAbelianInvariants(S[Length(S)])
+                            and Length(AbelianInvariants(S[Length(S)])) > 0)
+      ) do
```

Comment:
Ugh, this makes my head hurt... :). But it seems valid, and the code it replaces is not nicer, so no fundamental objection. Simplifying that doesn't seem easy... Hmmm

Perhaps we should add some functions like these:
```gap
IsAttributeKnownAndEqualTo := function(obj, attribute, value)
  return Tester(attribute)(obj) and attribute(obj) = value;
end;
IsKnownToBeTrivial := obj -> IsAttributeKnownAndEqualTo(obj, IsTrivial, true);
```
Then the code here could be changed to this (I made another change to avoid writing `S[Length(S)]` all the time; I didn't check this code, so I might have messed that up. But I hope my intent is clear):
```gap
S :=[];
D := G;
while not IsKnownToBeTrivial(D) and
         not IsAttributeKnownAndEqualTo(D, IsPerfectGroup, true) and
         not IsAttributeKnownAndEqualTo(D, AbelianInvariants, [] do
  Add( S, D );
  Info( InfoGroup, 2, "DerivedSeriesOfGroup: step ", Length(S) );
  D := DerivedSubgroup( D );
od;
Add( S, D );
```
I think there are other places in the GAP library which would become more readable with `IsAttributeKnownAndEqualTo`